fleetconfig-controller/docs/2-phase-spoke-reconcile.md

```diff
@@ -0,0 +1,227 @@
+3. Remove `HubCleanupFinalizer`
+
+In total, the cleanup process is completed in 3 reconciles - 2 on the hub and 1 on the spoke.
+
+**Special Cases**:
+
+There are 2 special cases to consider:
+
+- **hub-as-spoke**: The Spoke Reconciler (Hub) will also run the spoke-side cleanup steps right after the hub pre-flight cleanup steps. Cleanup completes in 2 reconciles.
+- **Failed Pivot**: The spoke agent never came up, so the hub will attempt the spoke-side cleanup steps right after the hub pre-flight cleanup steps. Cleanup completes in 2 reconciles.
+This allows for an "escape hatch" if the spoke agent never came up. This is the only case where a hub will perform day 2 operations on a spoke. It will never attempt upgrades on a spoke.
+
+## Key Points
+
+- All configuration is done on the hub cluster by CRUD operation on Spoke resources.
+- After the initial join, OCM addon framework is leveraged to install a fleetconfig-controller-agent inside the spoke cluster. The agent is responsible for the spoke's day 2 operations, including klusterlet upgrades and local cleanup.
+- After the "pivot", the spoke kubeconfig secret can be safely deleted. The hub will no longer directly manage the spoke cluster. Instead, the agent will asynchronously pull updates from the hub and reconcile them locally.
+- Leveraging finalizers to coordinate cleanup tasks allows the controllers to operate independently and avoid direct communication. Otherwise, API calls between the manager and agent would be required to coordinate cleanup.
+
+## Sequence Diagrams
+
+### Spoke Reconciles
+
+#### Actors:
+- User: End user
+- HubK8s: Hub cluster Kubernetes API server
+- Spoke Reconciler (Hub): Hub-side instance of the fleetconfig-controller-manager SpokeReconciler
+- SpokeK8s: Spoke cluster Kubernetes API server
+- Spoke Reconciler (Spoke): Spoke-side instance of the fleetconfig-controller-agent SpokeReconciler
+- Klusterlet CR: Klusterlet resource in the spoke cluster
+- Klusterlet Controllers: Klusterlet Controllers installed on the spoke cluster
+
+#### Day 1 - Join
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant HubK8s as Hub Cluster API Server
+    participant HubController as Spoke Reconciler (Hub)
+    participant SpokeK8s as Spoke Cluster API Server
+    participant KlusterletCtrl as Klusterlet Controllers
+
+    Note over User, HubController: Initialization
+
+    User->>HubK8s: Create Spoke resource
+    HubK8s->>HubController: Spoke resource created
+    HubController->>HubK8s: Add HubCleanupPreflightFinalizer
+    HubController->>HubK8s: Add HubCleanupFinalizer
+    HubController->>HubK8s: Add SpokeCleanupFinalizer (hub-as-spoke only)
+    
+    Note over HubController, KlusterletCtrl: Join Process
+    HubController->>HubController: Check if ManagedCluster exists
+    alt ManagedCluster does not exist
+        HubController->>SpokeK8s: Create Klusterlet and controllers (clusteradm join)
+        KlusterletCtrl->>HubK8s: Create CSR
```

Comment:
It's technically the registration agent, which is separate from the klusterlet controller :/